File: NewFramework/Utils/PlaywrightUtils.ts
import { Page, Locator, expect, BrowserContext } from '@playwright/test';
import { Logger } from './Logger';
import { ConfigManager } from './ConfigManager';
import * as path from 'path';
import * as fs from 'fs';

export class PlaywrightUtils {
  private page: Page;
  private logger: Logger;
  private config: ConfigManager;

  constructor(page: Page) {
    this.page = page;
    this.logger = Logger.getInstance();
    this.config = ConfigManager.getInstance();
  }

  // Navigation Methods
  async navigateTo(url: string): Promise<void> {
    this.logger.logAction(`Navigate to: ${url}`);
    await this.page.goto(url);
    await this.page.waitForLoadState('networkidle');
  }

  async goBack(): Promise<void> {
    this.logger.logAction('Navigate back');
    await this.page.goBack();
  }

  async goForward(): Promise<void> {
    this.logger.logAction('Navigate forward');
    await this.page.goForward();
  }

  async refresh(): Promise<void> {
    this.logger.logAction('Refresh page');
    await this.page.reload();
  }

  // Element Interaction Methods
  async click(selector: string, options?: { timeout?: number; force?: boolean }): Promise<void> {
    this.logger.logAction('Click', selector);
    await this.page.locator(selector).click(options);
  }

  async doubleClick(selector: string): Promise<void> {
    this.logger.logAction('Double click', selector);
    await this.page.locator(selector).dblclick();
  }

  async rightClick(selector: string): Promise<void> {
    this.logger.logAction('Right click', selector);
    await this.page.locator(selector).click({ button: 'right' });
  }

  async type(selector: string, text: string, options?: { delay?: number }): Promise<void> {
    this.logger.logAction(`Type '${text}'`, selector);
    await this.page.locator(selector).fill(''); // Clear first
    await this.page.locator(selector).type(text, options);
  }

  async fill(selector: string, text: string): Promise<void> {
    this.logger.logAction(`Fill '${text}'`, selector);
    await this.page.locator(selector).fill(text);
  }

  async clear(selector: string): Promise<void> {
    this.logger.logAction('Clear', selector);
    await this.page.locator(selector).fill('');
  }

  async pressKey(key: string): Promise<void> {
    this.logger.logAction(`Press key: ${key}`);
    await this.page.keyboard.press(key);
  }

  async pressKeyOnElement(selector: string, key: string): Promise<void> {
    this.logger.logAction(`Press key '${key}'`, selector);
    await this.page.locator(selector).press(key);
  }

  // Dropdown and Select Methods
  async selectByValue(selector: string, value: string): Promise<void> {
    this.logger.logAction(`Select by value '${value}'`, selector);
    await this.page.locator(selector).selectOption({ value });
  }

  async selectByText(selector: string, text: string): Promise<void> {
    this.logger.logAction(`Select by text '${text}'`, selector);
    await this.page.locator(selector).selectOption({ label: text });
  }

  async selectByIndex(selector: string, index: number): Promise<void> {
    this.logger.logAction(`Select by index ${index}`, selector);
    await this.page.locator(selector).selectOption({ index });
  }

  // Checkbox and Radio Methods
  async check(selector: string): Promise<void> {
    this.logger.logAction('Check', selector);
    await this.page.locator(selector).check();
  }

  async uncheck(selector: string): Promise<void> {
    this.logger.logAction('Uncheck', selector);
    await this.page.locator(selector).uncheck();
  }

  async isChecked(selector: string): Promise<boolean> {
    return await this.page.locator(selector).isChecked();
  }

  // Wait Methods
  async waitForElement(selector: string, timeout?: number): Promise<void> {
    this.logger.logAction('Wait for element', selector);
    await this.page.locator(selector).waitFor({ 
      timeout: timeout || this.config.getTestSettings().DefaultTimeout 
    });
  }

  async waitForElementToBeVisible(selector: string, timeout?: number): Promise<void> {
    this.logger.logAction('Wait for element to be visible', selector);
    await this.page.locator(selector).waitFor({ 
      state: 'visible',
      timeout: timeout || this.config.getTestSettings().DefaultTimeout 
    });
  }

  async waitForElementToBeHidden(selector: string, timeout?: number): Promise<void> {
    this.logger.logAction('Wait for element to be hidden', selector);
    await this.page.locator(selector).waitFor({ 
      state: 'hidden',
      timeout: timeout || this.config.getTestSettings().DefaultTimeout 
    });
  }

  async waitForText(selector: string, text: string, timeout?: number): Promise<void> {
    this.logger.logAction(`Wait for text '${text}'`, selector);
    await this.page.locator(selector).filter({ hasText: text }).waitFor({
      timeout: timeout || this.config.getTestSettings().DefaultTimeout
    });
  }

  async waitForUrl(url: string | RegExp, timeout?: number): Promise<void> {
    this.logger.logAction(`Wait for URL: ${url}`);
    await this.page.waitForURL(url, { 
      timeout: timeout || this.config.getTestSettings().DefaultTimeout 
    });
  }

  async waitForLoadState(state: 'load' | 'domcontentloaded' | 'networkidle' = 'networkidle'): Promise<void> {
    this.logger.logAction(`Wait for load state: ${state}`);
    await this.page.waitForLoadState(state);
  }

  async sleep(milliseconds: number): Promise<void> {
    this.logger.logAction(`Sleep for ${milliseconds}ms`);
    await this.page.waitForTimeout(milliseconds);
  }

  // Get Methods
  async getText(selector: string): Promise<string> {
    const text = await this.page.locator(selector).textContent() || '';
    this.logger.logAction(`Get text: '${text}'`, selector);
    return text;
  }

  async getValue(selector: string): Promise<string> {
    const value = await this.page.locator(selector).inputValue();
    this.logger.logAction(`Get value: '${value}'`, selector);
    return value;
  }

  async getAttribute(selector: string, attribute: string): Promise<string | null> {
    const value = await this.page.locator(selector).getAttribute(attribute);
    this.logger.logAction(`Get attribute '${attribute}': '${value}'`, selector);
    return value;
  }

  async getTitle(): Promise<string> {
    const title = await this.page.title();
    this.logger.logAction(`Get page title: '${title}'`);
    return title;
  }

  async getCurrentUrl(): Promise<string> {
    const url = this.page.url();
    this.logger.logAction(`Get current URL: '${url}'`);
    return url;
  }

  // Element State Methods
  async isVisible(selector: string): Promise<boolean> {
    return await this.page.locator(selector).isVisible();
  }

  async isHidden(selector: string): Promise<boolean> {
    return await this.page.locator(selector).isHidden();
  }

  async isEnabled(selector: string): Promise<boolean> {
    return await this.page.locator(selector).isEnabled();
  }

  async isDisabled(selector: string): Promise<boolean> {
    return await this.page.locator(selector).isDisabled();
  }

  async getElementCount(selector: string): Promise<number> {
    const count = await this.page.locator(selector).count();
    this.logger.logAction(`Element count: ${count}`, selector);
    return count;
  }

  // File Upload Methods
  async uploadFile(selector: string, filePath: string): Promise<void> {
    this.logger.logAction(`Upload file: ${filePath}`, selector);
    await this.page.locator(selector).setInputFiles(filePath);
  }

  async uploadMultipleFiles(selector: string, filePaths: string[]): Promise<void> {
    this.logger.logAction(`Upload multiple files: ${filePaths.join(', ')}`, selector);
    await this.page.locator(selector).setInputFiles(filePaths);
  }

  // Screenshot Methods
  async takeScreenshot(fileName?: string): Promise<string> {
    const screenshotDir = './Reports/screenshots';
    if (!fs.existsSync(screenshotDir)) {
      fs.mkdirSync(screenshotDir, { recursive: true });
    }
    
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const screenshotName = fileName || `screenshot-${timestamp}.png`;
    const screenshotPath = path.join(screenshotDir, screenshotName);
    
    await this.page.screenshot({ path: screenshotPath, fullPage: true });
    this.logger.logAction(`Screenshot saved: ${screenshotPath}`);
    return screenshotPath;
  }

  async takeElementScreenshot(selector: string, fileName?: string): Promise<string> {
    const screenshotDir = './Reports/screenshots';
    if (!fs.existsSync(screenshotDir)) {
      fs.mkdirSync(screenshotDir, { recursive: true });
    }
    
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const screenshotName = fileName || `element-screenshot-${timestamp}.png`;
    const screenshotPath = path.join(screenshotDir, screenshotName);
    
    await this.page.locator(selector).screenshot({ path: screenshotPath });
    this.logger.logAction(`Element screenshot saved: ${screenshotPath}`, selector);
    return screenshotPath;
  }

  // Assertion Helper Methods
  async assertElementVisible(selector: string, message?: string): Promise<void> {
    this.logger.logAction('Assert element visible', selector);
    await expect(this.page.locator(selector), message).toBeVisible();
  }

  async assertElementHidden(selector: string, message?: string): Promise<void> {
    this.logger.logAction('Assert element hidden', selector);
    await expect(this.page.locator(selector), message).toBeHidden();
  }

  async assertElementEnabled(selector: string, message?: string): Promise<void> {
    this.logger.logAction('Assert element enabled', selector);
    await expect(this.page.locator(selector), message).toBeEnabled();
  }

  async assertElementDisabled(selector: string, message?: string): Promise<void> {
    this.logger.logAction('Assert element disabled', selector);
    await expect(this.page.locator(selector), message).toBeDisabled();
  }

  async assertText(selector: string, expectedText: string, message?: string): Promise<void> {
    this.logger.logAction(`Assert text equals '${expectedText}'`, selector);
    await expect(this.page.locator(selector), message).toHaveText(expectedText);
  }

  async assertTextContains(selector: string, expectedText: string, message?: string): Promise<void> {
    this.logger.logAction(`Assert text contains '${expectedText}'`, selector);
    await expect(this.page.locator(selector), message).toContainText(expectedText);
  }

  async assertValue(selector: string, expectedValue: string, message?: string): Promise<void> {
    this.logger.logAction(`Assert value equals '${expectedValue}'`, selector);
    await expect(this.page.locator(selector), message).toHaveValue(expectedValue);
  }

  async assertTitle(expectedTitle: string, message?: string): Promise<void> {
    this.logger.logAction(`Assert page title equals '${expectedTitle}'`);
    await expect(this.page, message).toHaveTitle(expectedTitle);
  }

  async assertUrl(expectedUrl: string | RegExp, message?: string): Promise<void> {
    this.logger.logAction(`Assert URL equals '${expectedUrl}'`);
    await expect(this.page, message).toHaveURL(expectedUrl);
  }

  async assertElementCount(selector: string, expectedCount: number, message?: string): Promise<void> {
    this.logger.logAction(`Assert element count equals ${expectedCount}`, selector);
    await expect(this.page.locator(selector), message).toHaveCount(expectedCount);
  }

  // Scroll Methods
  async scrollToElement(selector: string): Promise<void> {
    this.logger.logAction('Scroll to element', selector);
    await this.page.locator(selector).scrollIntoViewIfNeeded();
  }

  async scrollToTop(): Promise<void> {
    this.logger.logAction('Scroll to top');
    await this.page.evaluate(() => window.scrollTo(0, 0));
  }

  async scrollToBottom(): Promise<void> {
    this.logger.logAction('Scroll to bottom');
    await this.page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
  }

  // Drag and Drop
  async dragAndDrop(sourceSelector: string, targetSelector: string): Promise<void> {
    this.logger.logAction(`Drag and drop from '${sourceSelector}' to '${targetSelector}'`);
    await this.page.locator(sourceSelector).dragTo(this.page.locator(targetSelector));
  }

  // Hover Methods
  async hover(selector: string): Promise<void> {
    this.logger.logAction('Hover', selector);
    await this.page.locator(selector).hover();
  }

  // Focus Methods
  async focus(selector: string): Promise<void> {
    this.logger.logAction('Focus', selector);
    await this.page.locator(selector).focus();
  }

  // Alert/Dialog Methods
  async acceptAlert(): Promise<void> {
    this.logger.logAction('Accept alert');
    this.page.on('dialog', dialog => dialog.accept());
  }

  async dismissAlert(): Promise<void> {
    this.logger.logAction('Dismiss alert');
    this.page.on('dialog', dialog => dialog.dismiss());
  }

  async getAlertText(): Promise<string> {
    return new Promise((resolve) => {
      this.page.on('dialog', dialog => {
        const message = dialog.message();
        this.logger.logAction(`Alert text: '${message}'`);
        resolve(message);
        dialog.accept();
      });
    });
  }

  // Frame Methods
  async switchToFrame(frameSelector: string): Promise<void> {
    this.logger.logAction('Switch to frame', frameSelector);
    const frame = this.page.frameLocator(frameSelector);
    // Return frame for chaining operations
    return frame as any;
  }

  // Cookie Methods
  async addCookie(name: string, value: string, domain?: string): Promise<void> {
    this.logger.logAction(`Add cookie: ${name}=${value}`);
    await this.page.context().addCookies([{
      name,
      value,
      domain: domain || new URL(this.page.url()).hostname,
      path: '/'
    }]);
  }

  async getCookies(): Promise<any[]> {
    const cookies = await this.page.context().cookies();
    this.logger.logAction(`Retrieved ${cookies.length} cookies`);
    return cookies;
  }

  async clearCookies(): Promise<void> {
    this.logger.logAction('Clear all cookies');
    await this.page.context().clearCookies();
  }

  // Local Storage Methods
  async setLocalStorage(key: string, value: string): Promise<void> {
    this.logger.logAction(`Set localStorage: ${key}=${value}`);
    await this.page.evaluate(([k, v]) => localStorage.setItem(k, v), [key, value]);
  }

  async getLocalStorage(key: string): Promise<string | null> {
    const value = await this.page.evaluate(k => localStorage.getItem(k), key);
    this.logger.logAction(`Get localStorage '${key}': '${value}'`);
    return value;
  }

  async clearLocalStorage(): Promise<void> {
    this.logger.logAction('Clear localStorage');
    await this.page.evaluate(() => localStorage.clear());
  }
}

File: NewFramework/CSharpTests/Utils/PlaywrightUtils.cs
using Microsoft.Playwright;
using System.Text.RegularExpressions;

namespace PlaywrightFramework.Utils
{
    public class PlaywrightUtils
    {
        private readonly IPage _page;
        private readonly Logger _logger;
        private readonly ConfigManager _config;

        public PlaywrightUtils(IPage page)
        {
            _page = page;
            _logger = Logger.Instance;
            _config = ConfigManager.Instance;
        }

        // Navigation Methods
        public async Task NavigateToAsync(string url)
        {
            _logger.LogAction($"Navigate to: {url}");
            await _page.GotoAsync(url);
            await _page.WaitForLoadStateAsync(LoadState.NetworkIdle);
        }

        public async Task GoBackAsync()
        {
            _logger.LogAction("Navigate back");
            await _page.GoBackAsync();
        }

        public async Task GoForwardAsync()
        {
            _logger.LogAction("Navigate forward");
            await _page.GoForwardAsync();
        }

        public async Task RefreshAsync()
        {
            _logger.LogAction("Refresh page");
            await _page.ReloadAsync();
        }

        // Element Interaction Methods
        public async Task ClickAsync(string selector, PageClickOptions? options = null)
        {
            _logger.LogAction("Click", selector);
            await _page.ClickAsync(selector, options);
        }

        public async Task DoubleClickAsync(string selector)
        {
            _logger.LogAction("Double click", selector);
            await _page.DblClickAsync(selector);
        }

        public async Task RightClickAsync(string selector)
        {
            _logger.LogAction("Right click", selector);
            await _page.ClickAsync(selector, new PageClickOptions { Button = MouseButton.Right });
        }

        public async Task TypeAsync(string selector, string text, PageTypeOptions? options = null)
        {
            _logger.LogAction($"Type '{text}'", selector);
            await _page.FillAsync(selector, ""); // Clear first
            await _page.TypeAsync(selector, text, options);
        }

        public async Task FillAsync(string selector, string text)
        {
            _logger.LogAction($"Fill '{text}'", selector);
            await _page.FillAsync(selector, text);
        }

        public async Task ClearAsync(string selector)
        {
            _logger.LogAction("Clear", selector);
            await _page.FillAsync(selector, "");
        }

        public async Task PressKeyAsync(string key)
        {
            _logger.LogAction($"Press key: {key}");
            await _page.Keyboard.PressAsync(key);
        }

        public async Task PressKeyOnElementAsync(string selector, string key)
        {
            _logger.LogAction($"Press key '{key}'", selector);
            await _page.PressAsync(selector, key);
        }

        // Dropdown and Select Methods
        public async Task SelectByValueAsync(string selector, string value)
        {
            _logger.LogAction($"Select by value '{value}'", selector);
            await _page.SelectOptionAsync(selector, new SelectOptionValue { Value = value });
        }

        public async Task SelectByTextAsync(string selector, string text)
        {
            _logger.LogAction($"Select by text '{text}'", selector);
            await _page.SelectOptionAsync(selector, new SelectOptionValue { Label = text });
        }

        public async Task SelectByIndexAsync(string selector, int index)
        {
            _logger.LogAction($"Select by index {index}", selector);
            await _page.SelectOptionAsync(selector, new SelectOptionValue { Index = index });
        }

        // Checkbox and Radio Methods
        public async Task CheckAsync(string selector)
        {
            _logger.LogAction("Check", selector);
            await _page.CheckAsync(selector);
        }

        public async Task UncheckAsync(string selector)
        {
            _logger.LogAction("Uncheck", selector);
            await _page.UncheckAsync(selector);
        }

        public async Task<bool> IsCheckedAsync(string selector)
        {
            return await _page.IsCheckedAsync(selector);
        }

        // Wait Methods
        public async Task WaitForElementAsync(string selector, int? timeout = null)
        {
            _logger.LogAction("Wait for element", selector);
            await _page.WaitForSelectorAsync(selector, new PageWaitForSelectorOptions 
            { 
                Timeout = timeout ?? _config.GetTestSettings().DefaultTimeout 
            });
        }

        public async Task WaitForElementToBeVisibleAsync(string selector, int? timeout = null)
        {
            _logger.LogAction("Wait for element to be visible", selector);
            await _page.WaitForSelectorAsync(selector, new PageWaitForSelectorOptions 
            { 
                State = WaitForSelectorState.Visible,
                Timeout = timeout ?? _config.GetTestSettings().DefaultTimeout 
            });
        }

        public async Task WaitForElementToBeHiddenAsync(string selector, int? timeout = null)
        {
            _logger.LogAction("Wait for element to be hidden", selector);
            await _page.WaitForSelectorAsync(selector, new PageWaitForSelectorOptions 
            { 
                State = WaitForSelectorState.Hidden,
                Timeout = timeout ?? _config.GetTestSettings().DefaultTimeout 
            });
        }

        public async Task WaitForTextAsync(string selector, string text, int? timeout = null)
        {
            _logger.LogAction($"Wait for text '{text}'", selector);
            await _page.Locator(selector).Filter(new LocatorFilterOptions { HasText = text })
                .WaitForAsync(new LocatorWaitForOptions 
                { 
                    Timeout = timeout ?? _config.GetTestSettings().DefaultTimeout 
                });
        }

        public async Task WaitForUrlAsync(string url, int? timeout = null)
        {
            _logger.LogAction($"Wait for URL: {url}");
            await _page.WaitForURLAsync(url, new PageWaitForURLOptions 
            { 
                Timeout = timeout ?? _config.GetTestSettings().DefaultTimeout 
            });
        }

        public async Task WaitForUrlAsync(Regex urlPattern, int? timeout = null)
        {
            _logger.LogAction($"Wait for URL pattern: {urlPattern}");
            await _page.WaitForURLAsync(urlPattern, new PageWaitForURLOptions 
            { 
                Timeout = timeout ?? _config.GetTestSettings().DefaultTimeout 
            });
        }

        public async Task WaitForLoadStateAsync(LoadState state = LoadState.NetworkIdle)
        {
            _logger.LogAction($"Wait for load state: {state}");
            await _page.WaitForLoadStateAsync(state);
        }

        public async Task SleepAsync(int milliseconds)
        {
            _logger.LogAction($"Sleep for {milliseconds}ms");
            await _page.WaitForTimeoutAsync(milliseconds);
        }

        // Get Methods
        public async Task<string> GetTextAsync(string selector)
        {
            var text = await _page.TextContentAsync(selector) ?? "";
            _logger.LogAction($"Get text: '{text}'", selector);
            return text;
        }

        public async Task<string> GetValueAsync(string selector)
        {
            var value = await _page.InputValueAsync(selector);
            _logger.LogAction($"Get value: '{value}'", selector);
            return value;
        }

        public async Task<string?> GetAttributeAsync(string selector, string attribute)
        {
            var value = await _page.GetAttributeAsync(selector, attribute);
            _logger.LogAction($"Get attribute '{attribute}': '{value}'", selector);
            return value;
        }

        public async Task<string> GetTitleAsync()
        {
            var title = await _page.TitleAsync();
            _logger.LogAction($"Get page title: '{title}'");
            return title;
        }

        public string GetCurrentUrl()
        {
            var url = _page.Url;
            _logger.LogAction($"Get current URL: '{url}'");
            return url;
        }

        public async Task<string> GetCurrentUrlAsync()
        {
            var url = _page.Url;
            _logger.LogAction($"Get current URL: '{url}'");
            return url;
        }

        // Element State Methods
        public async Task<bool> IsVisibleAsync(string selector)
        {
            return await _page.IsVisibleAsync(selector);
        }

        public async Task<bool> IsHiddenAsync(string selector)
        {
            return await _page.IsHiddenAsync(selector);
        }

        public async Task<bool> IsEnabledAsync(string selector)
        {
            return await _page.IsEnabledAsync(selector);
        }

        public async Task<bool> IsDisabledAsync(string selector)
        {
            return await _page.IsDisabledAsync(selector);
        }

        public async Task<int> GetElementCountAsync(string selector)
        {
            var count = await _page.Locator(selector).CountAsync();
            _logger.LogAction($"Element count: {count}", selector);
            return count;
        }

        // File Upload Methods
        public async Task UploadFileAsync(string selector, string filePath)
        {
            _logger.LogAction($"Upload file: {filePath}", selector);
            await _page.SetInputFilesAsync(selector, filePath);
        }

        public async Task UploadMultipleFilesAsync(string selector, string[] filePaths)
        {
            _logger.LogAction($"Upload multiple files: {string.Join(", ", filePaths)}", selector);
            await _page.SetInputFilesAsync(selector, filePaths);
        }

        // Screenshot Methods
        public async Task<string> TakeScreenshotAsync(string? fileName = null)
        {
            var screenshotDir = "./Reports/screenshots";
            if (!Directory.Exists(screenshotDir))
            {
                Directory.CreateDirectory(screenshotDir);
            }
            
            var timestamp = DateTime.Now.ToString("yyyy-MM-dd-HH-mm-ss");
            var screenshotName = fileName ?? $"screenshot-{timestamp}.png";
            var screenshotPath = Path.Combine(screenshotDir, screenshotName);
            
            await _page.ScreenshotAsync(new PageScreenshotOptions { Path = screenshotPath, FullPage = true });
            _logger.LogAction($"Screenshot saved: {screenshotPath}");
            return screenshotPath;
        }

        public async Task<string> TakeElementScreenshotAsync(string selector, string? fileName = null)
        {
            var screenshotDir = "./Reports/screenshots";
            if (!Directory.Exists(screenshotDir))
            {
                Directory.CreateDirectory(screenshotDir);
            }
            
            var timestamp = DateTime.Now.ToString("yyyy-MM-dd-HH-mm-ss");
            var screenshotName = fileName ?? $"element-screenshot-{timestamp}.png";
            var screenshotPath = Path.Combine(screenshotDir, screenshotName);
            
            await _page.Locator(selector).ScreenshotAsync(new LocatorScreenshotOptions { Path = screenshotPath });
            _logger.LogAction($"Element screenshot saved: {screenshotPath}", selector);
            return screenshotPath;
        }

        // Assertion Helper Methods
        public async Task AssertElementVisibleAsync(string selector, string? message = null)
        {
            _logger.LogAction("Assert element visible", selector);
            await Assertions.Expect(_page.Locator(selector)).ToBeVisibleAsync();
        }

        public async Task AssertElementHiddenAsync(string selector, string? message = null)
        {
            _logger.LogAction("Assert element hidden", selector);
            await Assertions.Expect(_page.Locator(selector)).ToBeHiddenAsync();
        }

        public async Task AssertElementEnabledAsync(string selector, string? message = null)
        {
            _logger.LogAction("Assert element enabled", selector);
            await Assertions.Expect(_page.Locator(selector)).ToBeEnabledAsync();
        }

        public async Task AssertElementDisabledAsync(string selector, string? message = null)
        {
            _logger.LogAction("Assert element disabled", selector);
            await Assertions.Expect(_page.Locator(selector)).ToBeDisabledAsync();
        }

        public async Task AssertTextAsync(string selector, string expectedText, string? message = null)
        {
            _logger.LogAction($"Assert text equals '{expectedText}'", selector);
            await Assertions.Expect(_page.Locator(selector)).ToHaveTextAsync(expectedText);
        }

        public async Task AssertTextContainsAsync(string selector, string expectedText, string? message = null)
        {
            _logger.LogAction($"Assert text contains '{expectedText}'", selector);
            await Assertions.Expect(_page.Locator(selector)).ToContainTextAsync(expectedText);
        }

        public async Task AssertValueAsync(string selector, string expectedValue, string? message = null)
        {
            _logger.LogAction($"Assert value equals '{expectedValue}'", selector);
            await Assertions.Expect(_page.Locator(selector)).ToHaveValueAsync(expectedValue);
        }

        public async Task AssertTitleAsync(string expectedTitle, string? message = null)
        {
            _logger.LogAction($"Assert page title equals '{expectedTitle}'");
            await Assertions.Expect(_page).ToHaveTitleAsync(expectedTitle);
        }

        public async Task AssertTitleContainsAsync(string expectedText, string? message = null)
        {
            _logger.LogAction($"Assert page title contains '{expectedText}'");
            await Assertions.Expect(_page).ToHaveTitleAsync(new Regex($".*{Regex.Escape(expectedText)}.*"));
        }

        public async Task AssertUrlAsync(string expectedUrl, string? message = null)
        {
            _logger.LogAction($"Assert URL equals '{expectedUrl}'");
            await Assertions.Expect(_page).ToHaveURLAsync(expectedUrl);
        }

        public async Task AssertUrlAsync(Regex expectedUrlPattern, string? message = null)
        {
            _logger.LogAction($"Assert URL matches pattern '{expectedUrlPattern}'");
            await Assertions.Expect(_page).ToHaveURLAsync(expectedUrlPattern);
        }

        public async Task AssertElementCountAsync(string selector, int expectedCount, string? message = null)
        {
            _logger.LogAction($"Assert element count equals {expectedCount}", selector);
            await Assertions.Expect(_page.Locator(selector)).ToHaveCountAsync(expectedCount);
        }

        // Scroll Methods
        public async Task ScrollToElementAsync(string selector)
        {
            _logger.LogAction("Scroll to element", selector);
            await _page.Locator(selector).ScrollIntoViewIfNeededAsync();
        }

        public async Task ScrollToTopAsync()
        {
            _logger.LogAction("Scroll to top");
            await _page.EvaluateAsync("window.scrollTo(0, 0)");
        }

        public async Task ScrollToBottomAsync()
        {
            _logger.LogAction("Scroll to bottom");
            await _page.EvaluateAsync("window.scrollTo(0, document.body.scrollHeight)");
        }

        // Drag and Drop
        public async Task DragAndDropAsync(string sourceSelector, string targetSelector)
        {
            _logger.LogAction($"Drag and drop from '{sourceSelector}' to '{targetSelector}'");
            await _page.Locator(sourceSelector).DragToAsync(_page.Locator(targetSelector));
        }

        // Hover Methods
        public async Task HoverAsync(string selector)
        {
            _logger.LogAction("Hover", selector);
            await _page.HoverAsync(selector);
        }

        // Focus Methods
        public async Task FocusAsync(string selector)
        {
            _logger.LogAction("Focus", selector);
            await _page.FocusAsync(selector);
        }
    }
}